src/rpc/net.cpp

```diff
@@ -200,6 +200,9 @@ static RPCHelpMan getpeerinfo()
         ServiceFlags services{fStateStats ? statestats.their_services : ServiceFlags::NODE_NONE};
         obj.pushKV("services", strprintf("%016x", services));
         obj.pushKV("servicesnames", GetServicesNames(services));
+        if (fStateStats) {
```

Comment:
> Can you clarify how this commit (standalone) is a bug fix that needs to be backported to 24.x? (what will break if this isn't done).

Other already-backported v24 commits aren't bug fixes; some are doc updates and some are even arguably features. 

The getpeerinfo RPC has many fields and they were in the same order in the help and the output in the previous releases. I don't see a compelling reason to break that for these two fields for v24 and needlessly make the RPC more confusing.

Reply:
JSON objects are by [definition](https://www.rfc-editor.org/rfc/rfc8259.html)  collections of name/value pairs that are unordered.
Given that, why not just shuffle the order in the help instead, so that we don't need extra conditionals in the source code and can group things together that belong together?

Reply:
On my Debian and macOS machines, the CLI and its help returns the getpeerinfo fields in the same order, which the codebase handles in the same order as well. This is handy; I don't see a very compelling reason to change the API needlessly to be less helpful for those who read the help and run CLI commands.

Reply:
> can group things together that belong together?

It would be potentially nice for that `fStateStats` flag grouping to go away; it has been a source of bugs in the past.

Reply:
> to be less helpful for those who read the help and run CLI commands.

The point is that machines / parsers consuming the JSON do not, and should not care about ordering. Which means nothing is "broken" if the fields happen to change order, and modifying the code to preserve an order that isn't guaranteed by anything is pointless. We also aren't optimising our RPC output for humans reading JSON in a terminal.

> to change the API needlessly

The API isn't changing though. The ordering is not guaranteed, and is not part of any API. The fields returned remain exactly the same.

I agree with Martins comment, if you care about the ordering being the same, it would make more sense to change the docs.

Reply:
Our RPC interface has the dual constraint of providing an API consumed by machines and the CLI interface used by humans -- cf the RPC helps that provide at least one example of each. It is therefore preferable to maintain the order, particularly when there is no need to change it.

Diffing `src/rpc/net.cpp` between the previous releases and master shows that changing the help order instead of the field order would mean needless churn in the interface. Needless, because there is no compelling reason. It makes more sense to keep the help the same in the absence of a real need to change it. A change ought to be justified, not stability. This commit chooses stability.

Reply:
>  shows that changing the help order instead of the field order would mean needless churn in the interface.

The documentation is not an interface, and changing the order of fields listed within it does not change the RPC interface/API. Ideally, the docs would be auto-generated from the code, and then more than likely, the fields would just be ordered alphanumerically.

> Needless, because there is no compelling reason. It makes more sense to keep the help the same in the absence of a real need to change it. A change from release to release ought to be justified, not stability. This commit chooses stability.

I disagree that the solution to maintaining something (field ordering) that isn't guaranteed, or part of any API, is [to add more complexity too](https://github.com/bitcoin/bitcoin/pull/26109#discussion_r977743011), or change the code. Especially when you could change the documentation and acheive the same thing.

Reply:
The diff to maintain stability is essentially the same without needlessly changing field position for the user. Code serves users.

Reply:
(I would like to remove the `fStateStats` grouping completely for the next release.)

Reply:
> Ideally the fields would just be ordered alphanumerically

That would be of benefit to neither machine nor humans? Rather it would be needless churn and confusion for humans reading the docs and using the RPC via the command line.

At least some of the RPCs return some of their output in an order that makes it easier to understand for a human, i.e. getpeerinfo begins with id/address/network fields for each peer. This seems helpful.

Whether everyone really wants to willy-nilly reorder all the fields and help docs across the board to make life confusing for people who use the CLI or not, it's a little strange to go way out on that limb in reaction to a minimal fixup.

Reply:
> That would be of benefit to neither machine nor humans?

It would benefit both (and developers) by:
* providing an ordering that is no-longer arbitrary
* which can be tested/linted/enforced
* same mechanism can be used to produce the field output, and the help output.
* we can stop caring about / trying to enforce / fix completely arbitrary orderings, that are not enforced / tested / checked by anything.

Reply:
Machines don't care about ordering and humans can find ordering helpful.  

Such a large change won't be done for v24 (if at all) and the argument isn't relevant here.